kimchi/src/circuits/expr.rs

```diff
@@ -1864,14 +2082,96 @@ impl<F: FftField, Column: PartialEq + Copy, ChallengeTerm: Copy>
     pub fn evaluations<
         'a,
-        Challenge: Index<ChallengeTerm, Output = F>,
-        Environment: ColumnEnvironment<'a, F, ChallengeTerm, Challenge, Column = Column>,
+        Challenges: Index<ChallengeTerm, Output = F>,
+        Environment: ColumnEnvironment<'a, F, ChallengeTerm, Challenges, Column = Column>,
     >(
         &self,
         env: &Environment,
     ) -> Evaluations<F, D<F>> {
         self.evaluate_constants(env).evaluations(env)
     }
+
+    /// Return an iterator over the evaluations provided by `evaluate_constants`
+    pub fn evaluations_iter<
+        'a,
+        'b: 'a,
+        Challenges: Index<ChallengeTerm, Output = F>,
+        Environment: ColumnEnvironment<'a, F, ChallengeTerm, Challenges, Column = Column>,
+    >(
+        &self,
+        env: &'b Environment,
+    ) -> EvaluationsIter<'b, F, ChallengeTerm, Challenges, Column, Environment> {
+        EvaluationsIter::new(self.evaluate_constants(env), env)
+    }
+}
+
+#[derive(Debug)]
+/// An `Iterator` structure for Evaluations.  See also the `impl EvaluationsIter` block below
+/// for additional handy methods.  Note that we do not provide the `Index` trait, since we
+/// don't return by reference.
+pub struct EvaluationsIter<'a, F: FftField, ChallengeTerm, Challenges, Column, Environment> {
+    idx: usize,
+    expr: Expr<F, Column>,
+    env: &'a Environment,
+    cache: HashMap<CacheId, Evaluations<F, D<F>>>,
+    _phantom: (
+        std::marker::PhantomData<Challenges>,
+        std::marker::PhantomData<ChallengeTerm>,
+    ),
+}
+
+impl<
+        'a,
+        'b,
+        F: FftField,
+        ChallengeTerm: Copy,
+        Challenges: Index<ChallengeTerm, Output = F>,
+        Column: PartialEq + Copy,
+        Environment: ColumnEnvironment<'a, F, ChallengeTerm, Challenges, Column = Column>,
+    > std::iter::Iterator for EvaluationsIter<'b, F, ChallengeTerm, Challenges, Column, Environment>
+where
+    'b: 'a,
+{
+    type Item = F;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ret = value_(&self.expr, self.env, &mut self.cache, self.idx);
+        self.idx += 1;
+        ret
+    }
+}
+
+impl<
+        'a,
+        'b,
+        F: FftField,
+        ChallengeTerm: Copy,
+        Challenges: Index<ChallengeTerm, Output = F>,
+        Column: PartialEq + Copy,
+        Environment: ColumnEnvironment<'a, F, ChallengeTerm, Challenges, Column = Column>,
+    > EvaluationsIter<'b, F, ChallengeTerm, Challenges, Column, Environment>
+where
+    'b: 'a,
+{
+    pub fn new(expr: Expr<F, Column>, env: &'b Environment) -> Self {
+        Self {
+            idx: 0,
+            expr,
+            env: &env,
+            cache: HashMap::new(),
+            _phantom: (std::marker::PhantomData, std::marker::PhantomData),
+        }
+    }
+
+    /// Populate the internal cache (by evaluating at the current row).
+    pub fn populate_cache(&mut self) {
+        value_(&self.expr, self.env, &mut self.cache, self.idx);
+    }
```

Comment:
This can be used to populate the `self.cache` parameter by evaluating over a single row.  In my testing, this tripled memory usage and did not affect performance.  I keep the method here in case my testing was naive.

Reply:
Note that besides this function, `value_` is no longer used.  If we decide to remove this function, `value_` should too be removed.  `clippy` should warn, since `value_` is not exported.